guides/hack/05-functions/30-function-references.md

```diff
@@ -0,0 +1,184 @@
+  called, for they have no implementation. Invoking a hypothetical reference to
+  one would also be an error, so we simply don’t allow a reference to be
+  created.
+- In traits and other classes that are not marked `final`, you cannot use
+  `self::` in a reference. This is to avoid ambiguity and confusion around what
+  `self` actually refers to when the method is called (it depends on the static
+  context of a call, whereas function references can never receive the context
+  information because the target is only resolved once).
+- Also, `parent::` is never supported.
+
+## Generics
+
+If you wish to pass along explicit generic arguments, either as a hint to the
+type checker, or in the case of a function with reified type parameters when
+they are required, that is also supported:
+
+```
+i_am_erased<int, _>; // erased generics (note wildcard)
+i_am_reified<C<string>>; // a reified generic
+```
+
+- Keep in mind that generics, if any, still must be provided at the location
+  where the function reference is created, rather than where it is used/invoked.
+- The arity of the type argument list, if it is non-empty (i.e. not `<>`), must
+  match the declaration, just like for an ordinary function call.
+- The special wildcard specifier `_` may be provided in place of any or all
+  erased (non-reified) generic arguments if you want Hack to infer a type
+  automatically based on the type parameter’s constraints. Again this is the
+  same as for ordinary function calls.
+
+Example with erased generics:
+
+```
+function fizz<Ta as num, Tb>(Ta $a, Tb $b): mixed { ... }
+
+$x = fizz<int, string>; // OK
+$x(4, 'hello');
+$x(-1, false); // error!
+
+$y = fizz<>;
+$y(3.14, new C()); // also OK
+$y('yo', derp()); // error!
+
+// OK as well
+$z = fizz<int, _>;
+$z = fizz<_, string>;
+$z = fizz<_, _>;
+
+// these all have errors!
+fizz<_>;
+fizz<string, _>;
+fizz<string, int>;
+```
+
+Example with [reified generics](/hack/generics/reified-generics):
+
+```
+function buzz<reify T as arraykey>(T $x): mixed { ... }
+
+$w = buzz<int>; // OK
+$w(42);
+$w("goodbye"); // error!
+
+// these all have errors!
+buzz<>;
+buzz<_>;
+buzz<mixed>;
+```
+
+## Introspection
+
+Function references can be cached in APC (the name will be resolved again when
+they are retrieved) or passed to memoized functions. However, other
+serialization formats are not supported.
+
+Internally, function/method references are represented using special data types
+that are intended to be opaque. This means they cannot (or should not) be cast
+directly to a string or another type, or be accessed in any other way besides
+calling them.
+
+It is *sometimes* useful to determine what a function reference is pointing to,
+e.g. for use in logging messages. For these *limited* use cases, if you know
+enough about the input and output formats you expect, HHVM provides the
+following helpers (but note they are not well supported and may change):
```

Comment:
We don't need to waffle as much in the documentation. These are the only native APIs that do what they do, and the ones you _should_ use, but internally we have our own that wrap them and we recommend the latter instead. Maybe we can add an FB-only blurb that mentions this?

Reply:
Sounds reasonable, do you have a suggestion on what exactly to put here? I'm not sure how much we want to encourage/discourage people from using these. If they're unsupported and likely to change, we should probably mention that.

We can add a FB-only block, but we're only hiding those using CSS so shouldn't mention anything sensitive.